### PR TITLE
Network consolidation

### PR DESCRIPTION
**Pull request by csduarte**
#### Summary
This PR attempts to consolidate all network related logic by pulling network connection listening, app state listening, and websocket initialization/starting/stopping into one component.

### CODE REVIEW

**Comment by enahum** [comment]
@jarredwitt build failures

File: app/components/network_indicator/network_indicator.js [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    static contextTypes = {
+        intl: intlShape.isRequired,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        const {isLandscape, networkStatus} = prevState;
+
+        if (nextProps.isLandscape !== isLandscape && networkStatus) {
+            const navBar = this.getNavBarHeight(nextProps.isLandscape);
+            const top = new Animated.Value(navBar - HEIGHT);
```
Comment (enahum) [inline app/components/network_indicator/network_indicator.js]: do we need this animated value in the state? I'm inclining not to do that for performance reasons, maybe take a look at the slide_up_panel component

File: app/components/network_indicator/network_indicator.js [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+            };
+        }
+
+        return null;
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.isX = DeviceInfo.getModel().includes('iPhone X');
```
Comment (enahum) [inline app/components/network_indicator/network_indicator.js]: this is a constant now for this `IS_IPHONE_X` in `constants/device`

File: app/components/network_indicator/network_indicator.js [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+    componentDidMount() {
+        this.mounted = true;
+    }
+
+    componentDidUpdate(prevProps) {
+        const {webSocketStatus} = this.props;
+
+        if (this.props.isOnline) {
+            if (this.state.networkStatus && webSocketStatus === RequestStatus.STARTED && this.props.webSocketStatus === RequestStatus.SUCCESS) {
```
Comment (enahum) [inline app/components/network_indicator/network_indicator.js]: I think `webSocketStatus` and `this.props.webSocketStatus` here is the same one, isn't it?

File: app/components/network_indicator/network_indicator.js [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                // Show the connected animation only if we had a previous network status
+                this.connected();
+            } else if (webSocketStatus === RequestStatus.STARTED && this.props.webSocketStatus === RequestStatus.FAILURE && this.props.isConnecting) {
```
Comment (enahum) [inline app/components/network_indicator/network_indicator.js]: same here

File: app/components/network_indicator/network_indicator.js [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+    connect = () => {
+        // need to change this to use the non-listener function
+        checkConnection((result) => {
+            this.setLocalState({networkStatus: CONNECTING}, () => {
+                if (result) {
+                    this.props.actions.connection(true);
+                    this.initializeWebSocket();
+                } else {
+                    this.setLocalState({networkStatus: OFFLINE});
```
Comment (enahum) [inline app/components/network_indicator/network_indicator.js]: shouldn't we here set the connection as false as well as to close the WS?

File: app/components/network_indicator/network_indicator.js [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    };
+
+    static contextTypes = {
+        intl: intlShape.isRequired,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        const {isLandscape, networkStatus} = prevState;
+
+        if (nextProps.isLandscape !== isLandscape && networkStatus) {
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/components/network_indicator/network_indicator.js]: Actually, we might not even need `networkStatus` in the state either. Doesn't that just come from `this.props.websocketStatus` anyway? If that's the case, we could get rid of the component state entirely

File: app/components/network_indicator/network_indicator.js (hunk 1 above)
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Yeah that would be great, looking at that component now.

File: app/components/network_indicator/network_indicator.js (hunk 2 above)
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Done

File: app/components/network_indicator/network_indicator.js (hunk 3 above)
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Yep, you're right. Fixed.

File: app/components/network_indicator/network_indicator.js (hunk 4 above)
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Fixed.

File: app/components/network_indicator/network_indicator.js (hunk 5 above)
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Yes, fixed.

File: app/components/network_indicator/network_indicator.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Yeah, I think that should be the goal. I'm taking a look at the slide_up_panel to see how we can match it to remove the animation from the state and will try to remove this from there as well.

**Comment by jarredwitt** [comment]
@enahum @hmhealey pushed some fixes so that this PR should build. I'm working on incorporating the other changes now.

**Comment by jarredwitt** [comment]
@enahum @hmhealey refactored and removed the animation from the state. I'm still trying to solve one case, but wanted to get your eyes on these changes quickly.

File: app/components/network_indicator/network_indicator.js [hunk 7]
```diff
@@ -35,85 +36,88 @@ const {
...
-    }
-
     componentWillUnmount() {
         this.mounted = false;
+
+        const {closeWebSocket, stopPeriodicStatusUpdates} = this.props.actions;
+
+        this.networkListener.removeEventListener();
+
+        AppState.removeEventListener('change', this.handleAppStateChange);
```
[the remark this answers is not in the document]
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Good catch. Added back in.

File: app/components/network_indicator/network_indicator.js [hunk 8]
```diff
@@ -35,85 +36,88 @@ const {
...
-            this.setLocalState({navBar, top});
-        }
+    componentDidUpdate(prevProps) {
+        const {websocketStatus: previousWebsocketStatus} = prevProps;
+        const {websocketErrorCount, websocketStatus} = this.props;
 
-        if (nextProps.isOnline) {
-            if (this.state.network && webSocketStatus === RequestStatus.STARTED && nextProps.webSocketStatus === RequestStatus.SUCCESS) {
+        if (this.props.isOnline) {
+            if (previousWebsocketStatus === RequestStatus.STARTED && websocketStatus === RequestStatus.SUCCESS) {
```
[the remark this answers is not in the document]
Comment (jarredwitt) [reply app/components/network_indicator/network_indicator.js]: Didn't notice it, but going to test simulating a slow network.

**Comment by jarredwitt** [comment]
@enahum I didn't notice any issues with the connection bar showing at app launch. I also added in a retry after 30 seconds. Can obviously change that, but here's a screenshot. I didn't do a countdown in order to keep from adding the component state back in.

image

**Comment by enahum** [comment]
@jarredwitt is there anything else pending on this PR ?

**Comment by enahum** [comment]
@jarredwitt is there anything else pending on this PR ? also I'm not seeing where `this.connectionRetryTimeout` is being set

File: app/components/network_indicator/network_indicator.js [hunk 9]
```diff
@@ -163,58 +169,112 @@ export default class OfflineIndicator extends Component {
...
+        if (this.isConnected !== isConnected) {
+            this.isConnected = isConnected;
+            this.handleWebSocket(isConnected);
+            connection(isConnected);
+        }
+    };
+
+    handleReconnect = () => {
+        clearTimeout(this.connectionRetryTimeout);
+        this.connectionRetryTimeout = setTimeout(() => {
```
Comment (jarredwitt) [inline app/components/network_indicator/network_indicator.js]: @enahum `this.connectionRetryTimeout` is being set here

**Comment by jarredwitt** [comment]
@enahum I don't think there's anything else pending on my end. Other than a solid testing round, I think it should be good to go.

**Comment by enahum** [comment]
@jarredwitt ok lets merge it and see ;)

**Pull request merged**